File: src/main/java/jp/vmi/selenium/selenese/DialogOverride.java
package jp.vmi.selenium.selenese;

import java.util.Map;

import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

import com.thoughtworks.selenium.SeleniumException;

import jp.vmi.selenium.selenese.utils.JSFunction;

/**
 * Override dialog interfaces.
 */
public class DialogOverride {

    private static final JSFunction replaceAlertMethod;
    private static final JSFunction getNextAlert;
    private static final JSFunction isAlertPresent;
    private static final JSFunction getNextConfirmation;
    private static final JSFunction isConfirmationPresent;
    private static final JSFunction answerOnNextPrompt;
    private static final JSFunction getNextPrompt;
    private static final JSFunction isPromptPresent;

    static {
        Map<String, JSFunction> jsMap = JSFunction.load(DialogOverride.class.getResourceAsStream("DialogOverride.js"));
        replaceAlertMethod = jsMap.get("replaceAlertMethod");
        getNextAlert = jsMap.get("getNextAlert");
        isAlertPresent = jsMap.get("isAlertPresent");
        getNextConfirmation = jsMap.get("getNextConfirmation");
        isConfirmationPresent = jsMap.get("isConfirmationPresent");
        answerOnNextPrompt = jsMap.get("answerOnNextPrompt");
        getNextPrompt = jsMap.get("getNextPrompt");
        isPromptPresent = jsMap.get("isPromptPresent");
    }

    /**
     * Replace alert/confirm/prompt methods.
     *
     * @param driver WebDriver object.
     * @param element target element for detecting frame/iframe.
     */
    public void replaceAlertMethod(WebDriver driver, WebElement element) {
        replaceAlertMethod.call(driver, element);
    }

    /**
     * Get next alert dialog message.
     *
     * @param driver WebDriver object.
     * @return alert dialog message.
     */
    public String getNextAlert(WebDriver driver) {
        String result = getNextAlert.call(driver);
        if (result == null)
            throw new SeleniumException("There were no alerts");
        return result;
    }

    /**
     * Test of whether a alert dialog was displayed.
     *
     * @param driver WebDriver object.
     * @return alert dialog is dislayed if true.
     */
    public boolean isAlertPresent(WebDriver driver) {
        return Boolean.TRUE.equals(isAlertPresent.call(driver));
    }

    /**
     * Get next confirmation dialog message.
     *
     * @param driver WebDriver object.
     * @return confirmation dialog message.
     */
    public String getNextConfirmation(WebDriver driver) {
        String result = getNextConfirmation.call(driver);
        if (result == null)
            throw new SeleniumException("There were no confirmations");
        return result;
    }

    /**
     * Test of whether a confirmation dialog was displayed.
     *
     * @param driver WebDriver object.
     * @return confirmation dialog is dislayed if true.
     */
    public boolean isConfirmationPresent(WebDriver driver) {
        return Boolean.TRUE.equals(isConfirmationPresent.call(driver));
    }

    /**
     * Set answer message for prompt.
     *
     * @param driver WebDriver object.
     * @param msg answer message used on next prompt.
     */
    public void answerOnNextPrompt(WebDriver driver, String msg) {
        answerOnNextPrompt.call(driver, msg);
    }

    /**
     * Get next prompt.
     *
     * @param driver WebDriver object.
     * @return prompt.
     */
    public String getNextPrompt(WebDriver driver) {
        String result = getNextPrompt.call(driver);
        if (result == null)
            throw new SeleniumException("There were no prompts");
        return result;
    }

    /**
     * Test of whether a prompt was displayed.
     *
     * @param driver WebDriver object.
     * @return prompt is dislayed if true.
     */
    public boolean isPromptPresent(WebDriver driver) {
        return Boolean.TRUE.equals(isPromptPresent.call(driver));
    }
}

File: src/main/resources/jp/vmi/selenium/selenese/DialogOverride.js
function replaceAlertMethod(element) { // BEGIN
  window.localStorage.setItem('__webdriverAlerts', JSON.stringify([]));
  window.alert = function(msg) {
    var alerts = JSON.parse(window.localStorage.getItem('__webdriverAlerts'));
    alerts.push(msg);
    window.localStorage.setItem('__webdriverAlerts', JSON.stringify(alerts));
  };
  window.localStorage.setItem('__webdriverConfirms', JSON.stringify([]));
  if (!('__webdriverNextConfirm' in window.localStorage))
    window.localStorage.setItem('__webdriverNextConfirm', JSON.stringify(true));
  window.confirm = function(msg) {
    var confirms = JSON.parse(window.localStorage.getItem('__webdriverConfirms'));
    confirms.push(msg);
    window.localStorage.setItem('__webdriverConfirms', JSON.stringify(confirms));
    var res = JSON.parse(window.localStorage.getItem('__webdriverNextConfirm'));
    window.localStorage.setItem('__webdriverNextConfirm', JSON.stringify(true));
    return res;
  };
  window.localStorage.setItem('__webdriverPrompts', JSON.stringify([]));
  if (!('__webdriverNextPrompt' in window.localStorage))
    window.localStorage.setItem('__webdriverNextPrompt', JSON.stringify(""));
  window.prompt = function(msg) {
    var prompts = JSON.parse(window.localStorage.getItem('__webdriverPrompts'));
    prompts.push(msg);
    window.localStorage.setItem('__webdriverPrompts', JSON.stringify(prompts));
    var res = JSON.parse(window.localStorage.getItem('__webdriverNextPrompt'));
    window.localStorage.setItem('__webdriverNextPrompt', JSON.stringify(""));
    return res;
  };
  var fw;
  if (element && (fw = element.ownerDocument.defaultView) && fw != window) {
    fw.alert = window.alert;
    fw.confirm = window.confirm;
    fw.prompt = window.prompt;
  }
} // END

function getNextAlert() { // BEGIN
  if (!('__webdriverAlerts' in window.localStorage))
    return null;
  var alerts = JSON.parse(window.localStorage.getItem('__webdriverAlerts'));
  if (! alerts)
    return null;
  var msg = alerts.shift();
  window.localStorage.setItem('__webdriverAlerts', JSON.stringify(alerts));
  if (msg)
    msg = msg.replace(/\\n/g, ' ');
  return msg;
} // END

function isAlertPresent() { // BEGIN
  if (!('__webdriverAlerts' in window.localStorage))
    return false;
  var alerts = JSON.parse(window.localStorage.getItem('__webdriverAlerts'));
  return alerts && alerts.length > 0;
} // END

function getNextConfirmation() { // BEGIN
  if (!('__webdriverConfirms' in window.localStorage))
    return null;
  var confirms = JSON.parse(window.localStorage.getItem('__webdriverConfirms'));
  if (! confirms)
    return null;
  var msg = confirms.shift();
  window.localStorage.setItem('__webdriverConfirms', JSON.stringify(confirms));
  if (msg)
    msg = msg.replace(/\\n/g, ' ');
  return msg;
} // END

function isConfirmationPresent() { // BEGIN
  if (!('__webdriverConfirms' in window.localStorage))
    return false;
  var confirms = JSON.parse(window.localStorage.getItem('__webdriverConfirms'));
  return confirms && confirms.length > 0;
} // END

function answerOnNextPrompt(answer) { // BEGIN
  window.localStorage.setItem('__webdriverNextPrompt', JSON.stringify(answer));
} // END

function getNextPrompt() { // BEGIN
  if (!('__webdriverPrompts' in window.localStorage))
    return null;
  var prompts = JSON.parse(window.localStorage.getItem('__webdriverPrompts'));
  if (!prompts)
    return null;
  var msg = prompts.shift();
  window.localStorage.setItem('__webdriverPrompts', JSON.stringify(prompts));
  if (msg)
    msg = msg.replace(/\\n/g, ' ');
  return msg;
} // END

function isPromptPresent() { // BEGIN
  if (!('__webdriverPrompts' in window.localStorage))
    return false;
  var prompts = JSON.parse(window.localStorage.getItem('__webdriverPrompts'));
  return prompts && prompts.length > 0;
} // END
